### PR TITLE
feat: DialogInspector and new test actions

### LINKED ISSUE #2474: [PORT] add Microsoft.Test.SettingsPropertiesMock
> Port this change from botbuilder-dotnet/master branch:
https://github.com/microsoft/botbuilder-dotnet/pull/4201

## Description

Towards testing declarative bot with only declarative files.

Add Microsoft.Test.SettingsPropertiesMock for mocking settings. For example, when one commit the project, some settings will be left empty like 'connectionName', 'apiUrl' etc. With this mock, we could set these values for testing purpose.

## Specific Changes

  - Add PropertyMock as interface to enable loading from file.
  - Add Microsoft.Test.SettingsPropertiesMock which is like SetProperties in schema. Note it **only supports string** as value because `configBuilder.AddInMemoryCollection` only supports string.
  - Add MockSettingsMiddleware for the real mock setup. Assume IConfiguration is used in turnContext.TurnState for settings scope.

## Testing

Add TestScriptTests_PropertyMock.test.dialog for testing assignments, overwrites.

# Changed projects
* Microsoft.Bot.Builder.Dialogs.Adaptive.Testing
* Microsoft.Bot.Builder.Dialogs.Adaptive.Tests
* 

[Adaptive]

### LINKED ISSUE #3035: [PORT] Add AssertNoActivity
https://github.com/microsoft/botbuilder-dotnet/pull/4785/

Add AssertNoActivity test action.
Add description to MemoryAssertion.
Fix bug in SetProperties where it did not evaluate expression.

### LINKED ISSUE #3113: [PORT] Hide DialogInspector and centralize resetting retries
https://github.com/microsoft/botbuilder-dotnet/pull/4715

This hides DialogInspector behind a delegate to reduce surface area and moves resetting retries from memory modification to when events are emitted making it simpler and clearer.

### PR DESCRIPTION
**Pull request by chon219**
Fixes #3113 
Fixes #3035 
Fixes #2474 

## Description

Implemented `DialogInspector` with some new test actions.

## Specific Changes

  - Implemented `DialogInspector`
  - Added `AssertNoReply`
  - Added `MemoryAssertions`
  - Added `SetProperties` (Microsoft.Test.SetProperties)
  - Centralized resetting retries

## Testing

Added test `PropertyMock`.

### CODE REVIEW

**Comment by chon219** [comment]
#3113 should be fixed in this PR as well.

File: libraries/botbuilder-dialogs-adaptive-testing/src/dialogInspector.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+     * Gets initialTurnState collection to copy into the turnState on every turn.
+     */
+    public readonly initialTurnState = new TurnContextStateCollection();
+
+    /**
+     * Gets root dialog to use to start conversation.
+     *
+     * @returns {Dialog} The root dialog.
+     */
+    public get rootDialog(): Dialog {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive-testing/src/dialogInspector.ts:74]: If possible, add `| undefined` to the return type signature. In strict mode, `undefined` does not satisfy any types aside from `undefined`.

File: libraries/botbuilder-dialogs-adaptive-testing/src/testAction.ts [hunk 2]
```diff
@@ -8,7 +8,20 @@
...
+
+/**
+ * Abstract base class for scripted actions.
+ */
 export abstract class TestAction extends Configurable {
-    public abstract execute(adapter: TestAdapter, callback: (context: TurnContext) => Promise<void>): void;
+    public abstract execute(
+        adapter: TestAdapter,
+        callback: (context: TurnContext) => Promise<void>,
+        inspector: Inspector
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive-testing/src/testAction.ts:25]: Should `inspector` be optional? This is probably not a GA library, but adding a new non-optional argument to a method signature is breaking.

File: libraries/botbuilder-dialogs-adaptive-testing/src/testActions/memoryAssertions.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+     * @param callback Logic for the bot to use.
+     * @param inspector Inspector for dialog context.
+     * @returns A Promise that represents the work queued to execute.
+     */
+    public async execute(
+        adapter: TestAdapter,
+        callback: (context: TurnContext) => Promise<void>,
+        inspector?: Inspector
+    ): Promise<void> {
+        if (inspector) {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive-testing/src/testActions/memoryAssertions.ts:46]: Is it an error case to use this assertion without passing in an inspector? Perhaps an error should be thrown in an else block.

File: libraries/botbuilder-dialogs-adaptive-testing/src/testAction.ts (hunk 2 above)
Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive-testing/src/testAction.ts:25]: Yes, it should be optional.

File: libraries/botbuilder-dialogs-adaptive-testing/src/testActions/memoryAssertions.ts (hunk 3 above)
Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive-testing/src/testActions/memoryAssertions.ts:46]: In C#, there is a trace info in the else block:
`Trace.TraceInformation($"[Turn Ended => No inspector for {Description} MemoryAssertions]");`
I'm not sure if we should throw errors if there is no inspector although the memory assertion requires inspector to run.

**Review by joshgummersall — APPROVED** [review]

**Pull request merged**